File: 3558-find-a-safe-walk-through-a-grid/find-a-safe-walk-through-a-grid.go
package main

// Point stores the coordinates for the queue
type Point struct {
	r, c int
}

func findSafeWalk(grid [][]int, health int) bool {
	m := len(grid)
	n := len(grid[0])

	// Health remaining after entering the starting cell
	startHealth := health - grid[0][0]

	if startHealth <= 0 {
		return false
	}

	// Initialize the best matrix with -1
	best := make([][]int, m)
	for i := range best {
		best[i] = make([]int, n)
		for j := range best[i] {
			best[i][j] = -1
		}
	}
	best[0][0] = startHealth

	// Standard slice-based queue with a head pointer for efficiency
	q := []Point{{0, 0}}
	head := 0

	directions := []Point{
		{1, 0}, {-1, 0}, {0, 1}, {0, -1},
	}

	for head < len(q) {
		// Popleft equivalent
		curr := q[head]
		head++
		
		r, c := curr.r, curr.c
		currentHealth := best[r][c]

		if r == m-1 && c == n-1 {
			return true
		}

		for _, dir := range directions {
			nr, nc := r+dir.r, c+dir.c

			// Bounds check
			if nr < 0 || nr >= m || nc < 0 || nc >= n {
				continue
			}

			newHealth := currentHealth - grid[nr][nc]

			if newHealth <= 0 {
				continue
			}

			if newHealth > best[nr][nc] {
				best[nr][nc] = newHealth
				q = append(q, Point{nr, nc})
			}
		}
	}

	return false
}

File: 3558-find-a-safe-walk-through-a-grid/find-a-safe-walk-through-a-grid.py
from collections import deque

class Solution:
    def findSafeWalk(self, grid, health):
        m, n = len(grid), len(grid[0])

        # Health remaining after entering the starting cell
        start_health = health - grid[0][0]

        if start_health <= 0:
            return False

        # best[r][c] = maximum health remaining seen so far
        best = [[-1] * n for _ in range(m)]
        best[0][0] = start_health

        q = deque([(0, 0)])

        directions = [(1,0), (-1,0), (0,1), (0,-1)]

        while q:
            r, c = q.popleft()

            current_health = best[r][c]

            if r == m - 1 and c == n - 1:
                return True

            for dr, dc in directions:
                nr, nc = r + dr, c + dc

                if not (0 <= nr < m and 0 <= nc < n):
                    continue

                new_health = current_health - grid[nr][nc]

                if new_health <= 0:
                    continue

                if new_health > best[nr][nc]:
                    best[nr][nc] = new_health
                    q.append((nr, nc))

        return False